### PR TITLE
Copy string before returning it

### PR DESCRIPTION
**Pull request by etiotto**
This PR fixes issue #1209.

Signed-off-by: Ettore Tiotto <etiotto@ca.ibm.com>

### CODE REVIEW

File: src/Compiler/OnnxMlirCompiler.cpp [hunk 1]
```diff
@@ -42,7 +42,7 @@ ONNX_MLIR_EXPORT int64_t omSetCompilerOption(
 
 ONNX_MLIR_EXPORT const char *omGetCompilerOption(const OptionKind kind) {
   std::string val = getCompilerOption(kind);
-  return val.c_str();
+  return strdup(val.c_str());
```
[the remark this answers is not in the document]
Comment (etiotto) [reply src/Compiler/OnnxMlirCompiler.cpp:45]: > The options are all in static area? So theoretically we should be able to return a pointer to the static area without doing a strdup?

The local variable `val` will go out of scope so if it copied the option string into a local buffer the memory the `char *` pointer points to will be gone.

[the remark this answers is not in the document]
Comment (tungld) [reply src/Compiler/OnnxMlirCompiler.cpp:45]: Can callers free a `const char*`?

[the remark this answers is not in the document]
Comment (gongsu832) [reply src/Compiler/OnnxMlirCompiler.cpp:45]: > Can callers free a `const char*`?

You will get a compiler warning unless you cast away the `const` before calling `free`.

[the remark this answers is not in the document]
Comment (gongsu832) [reply src/Compiler/OnnxMlirCompiler.cpp:45]: Not questioning whether strdup is needed or not. Yes in the current code it's needed. There is a bigger question than strdup. If we stick to static options, then we need to fix `omSetCompilerOptions` to be thread-safe. If you look at `CompilerUtils.cpp` there is a TODO comment above the static opts. So my point was that the whole static opts thing is probably not done properly and may need some restructure.

[the remark this answers is not in the document]
Comment (etiotto) [reply src/Compiler/OnnxMlirCompiler.cpp:45]: Ok we agree on the need for a strdup in that function, good.  Given that MLIR and LLVM use the `cl::opt` facilities to handle compiler option IMO we should do use the same facilities, which leaves us with the thread-safety issue you mentioned.  To address that issue I am going to use a `std::mutex` in `setCompilerOption` with will prevent a data race.

[the remark this answers is not in the document]
Comment (gongsu832) [reply src/Compiler/OnnxMlirCompiler.cpp:45]: Can you put some comments on why we are adding the mutex?

[the remark this answers is not in the document]
Comment (etiotto) [reply src/Compiler/OnnxMlirCompiler.cpp:45]: done

**Comment by etiotto** [comment]
@gongsu832 @AlexandreEichenberger can you pls take a look again ?

**Review by AlexandreEichenberger — CHANGES_REQUESTED** [review]

**Comment by gongsu832** [comment]
I respectfully disagree with removing the mutex. It's not to give a false sense of usability. Any time when you have multiple threads writing to shared data, serialization is the very basic requirement for not corrupting the shared data. The issue that one thread's changing the option will be seen by all other threads is a completely different issue and mutex is not even designed to protect against that.

**Comment by etiotto** [comment]
> So I am requesting that

> the mutex is removed for now because they are marginally useful and giving a false sense of usability.
> we label the function in the interface as to be called only in a single thread mode

I have removed the mutex because is not essential to the main goal of this patch.  There is disagreement on what to do w.r.t. thread safety as per discussion above.  I do think that having a mutex does not make things worse, however in order to make progress I think we can agree on having the `strdup` and leave the discussion on thread safety to another time :-)

**Review by AlexandreEichenberger — APPROVED** [review]
LGTM, agree mutex is tangential to this.

File: src/Compiler/CompilerUtils.cpp [hunk 2]
```diff
@@ -411,7 +412,12 @@ static std::string getLLVMOption() {
 // =============================================================================
 // Methods for OMCompilerOptions
 
+// Set the compiler options, use a mutex to allow multiple threads to set the
+// compiler options in a thread-safe manner.
 int setCompilerOption(const OptionKind kind, const string &val) {
+  static std::mutex mutex;
```
[the remark this answers is not in the document]
Comment (gongsu832) [reply src/Compiler/CompilerUtils.cpp]: Why is thread local ugly?

You are confusing inconsistent state (as in your -O3 and -O0 example) with corrupt data. Mutex does not prevent the inconsistent state in your example, that's known. Mutex prevents corrupt data as show in the example below:
```
static std::string getTargetArchOption() {
  return (march != "") ? "--march=" + march : "";
}
```
With proper mutex, the above code will always return either `--march=something` or empty string. Without mutex, the code could return a half baked `--march=` string, which may cause other code to fail. Yes in this particular case the half baked string may not be harmful but that's not the point.

**Pull request merged**